### ISSUE TITLE
Cookies and Tokens in same Web site

### ISSUE THREAD

**Issue by DaljitBhail** [openiddict/openiddict-core#343]
I'm still learning and developing Token and Cookie usage for an ASP.Net Core MVC Site.
I would want OpenIDDict to handle token authentication and authorization for my APIs and standard Cookie authentication and authorization for my views. I figured I would create Claim Policies bound to each authorization service i.e. all APIs authorization and authentication to OpenIDDict and View Pages to Cookie Authentication. Not sure having both on the same site is good practice but ensures underlying logic in model isn't duplicated across projects. 

I've got OpenIDDict working (after a week!) but struggling now with cookies. Although the cookie authentication works, when I have an "unauthorised" view it doesn't redirect to an unauthorised view. I'm not sure if my setup is correct or the best way to proceed. Any help would be appreciated as Ive struggled with this for weeks ??

```
using Deploy.Data;
using Deploy.Models;
using Deploy.Models.Blogs;
using Deploy.Services;
using Microsoft.AspNetCore.Builder;
using Microsoft.AspNetCore.Hosting;
using Microsoft.AspNetCore.Http;
using Microsoft.AspNetCore.Mvc;
using Microsoft.AspNetCore.Mvc.Cors.Internal;
using Microsoft.EntityFrameworkCore;
using Microsoft.Extensions.Configuration;
using Microsoft.Extensions.DependencyInjection;
using Microsoft.Extensions.Logging;
using Microsoft.IdentityModel.Tokens;
using System;
using System.Security.Claims;
using System.Security.Cryptography.X509Certificates;
using System.Text;

namespace Deploy
{
    public class Startup
    {
        public Startup(IHostingEnvironment env)
        {
            var builder = new ConfigurationBuilder()
                .SetBasePath(env.ContentRootPath)
                .AddJsonFile("appsettings.json", optional: true, reloadOnChange: true)
                .AddJsonFile($"appsettings.{env.EnvironmentName}.json", optional: true);

            if (env.IsDevelopment())
            {
                // For more details on using the user secret store see http://go.microsoft.com/fwlink/?LinkID=532709
                builder.AddUserSecrets("aspnet-Test_DEL-1d63ea04-757e-4d82-9b14-55eb8723c96a");
            }

            builder.AddEnvironmentVariables();
            Configuration = builder.Build();
        }

        public IConfigurationRoot Configuration { get; }

        // This method gets called by the runtime. Use this method to add services to the container.
        public void ConfigureServices(IServiceCollection services)
        {
            string cookieAuthenticationScheme = "";


            // Add framework services.
            services.AddDbContext<ApplicationDbContext>(options =>
            {
                options.UseSqlServer(Configuration.GetConnectionString("DefaultConnection"));
                options.UseOpenIddict<int>();
            }
            );

            #region Authentication
            //services.AddIdentity<ApplicationUser, ApplicationRole>()
            //    .AddEntityFrameworkStores<ApplicationDbContext, int>()
            //    .AddDefaultTokenProviders();

            //TODO isues with password. still insists on 6 chars
            services.AddIdentity<ApplicationUser, ApplicationRole>(options =>
            {
               
                // configure identity options
                options.Password.RequireDigit = false;
                options.Password.RequireLowercase = false;
                options.Password.RequireUppercase = false;
                options.Password.RequireNonAlphanumeric = false;
                options.Password.RequiredLength = 3;

                //  cookieAuthenticationScheme = options.Cookies.ExternalCookieAuthenticationScheme;
                //options.Cookies.ApplicationCookie.AutomaticAuthenticate = false; // These cause issues and shouldnt be spcidied at all
                //options.Cookies.ApplicationCookie.AutomaticChallenge = false;
                options.Cookies.ApplicationCookie.LoginPath = new PathString("/Account/Login");
                options.Cookies.ApplicationCookie.AccessDeniedPath = new PathString("/Account/AccessDenied");

                // User settings
                options.User.RequireUniqueEmail = true;
            })
                .AddEntityFrameworkStores<ApplicationDbContext, int>()
                .AddDefaultTokenProviders()
                ;

            services.Configure<CookieAuthenticationOptions>(options =>
            {
                options.LoginPath = new PathString("/Account/Login");
                options.AccessDeniedPath = new PathString("/Account/AccessDenied");
                options.AutomaticChallenge = false;
                options.AutomaticAuthenticate = false; //should be false
                //   cookieAuthenticationScheme = options.AuthenticationScheme;
                //  options.co
            });

            Console.Write("scheme name  is " + cookieAuthenticationScheme);

            var jwtSigningCert = new X509Certificate2("mycert.pfx", "password", X509KeyStorageFlags.MachineKeySet);
            services.AddOpenIddict<int>()
               .AddEntityFrameworkCoreStores<ApplicationDbContext>()
               .AddMvcBinders()
               .EnableTokenEndpoint("/connect/token")
               .AllowPasswordFlow()
               .AllowRefreshTokenFlow()
               .SetAccessTokenLifetime(TimeSpan.FromMinutes(10))
               .DisableHttpsRequirement()
               //.RequireClientIdentification()
               //.EnableRequestCaching()
               //   .AddEphemeralSigningKey()
               //.SetRefreshTokenLifetime(TimeSpan.FromMinutes(20))
               .UseJsonWebTokens() // mUst have JWT MIddleware to use this 
               .AddSigningCertificate(jwtSigningCert)
               ;


            #region Cors
            //allows Cors for API 
            //NB: must procede any endpoints
            services.AddCors(o => o.AddPolicy("AllowAll", builder =>
            {
                builder.AllowAnyOrigin()
                       .AllowAnyMethod()
                       .AllowAnyHeader()
                       .AllowCredentials()
                       .SetPreflightMaxAge(TimeSpan.FromSeconds(2520))
                       .Build();
            }));
            // Apply as default to all controllers. API etc
            services.Configure<MvcOptions>(options =>
            {
                options.Filters.Add(new CorsAuthorizationFilterFactory("AllowAll"));
            });
            #endregion
            #endregion


            //TODO need to redictecrt to no authroised for standard users
            #region Configure all Claims Policies
            services.AddAuthorization(options =>
            {
                //options.AddPolicy("Administrators", policy => policy.RequireRole("Admin"));
                //options.AddPolicy("Administrators", policy => policy.RequireClaim(ClaimTypes.Role, "Admin")); // This works the same as the above code

                options.AddPolicy("Administrators", policy =>
                {
                    //policy.AuthenticationSchemes.Add(cookieAuthenticationScheme);
                    policy.RequireAuthenticatedUser();
                    policy.RequireClaim(ClaimTypes.Role, "Admin");
                });

                //options.AddPolicy("Name", policy => policy.RequireClaim(ClaimTypes.Name, "Snotchie"));
                options.AddPolicy("Name", policy =>
                {
                    // policy.AuthenticationSchemes.Add("Cookie");
                    //policy.AuthenticationSchemes.Add(cookieAuthenticationScheme);
                    policy.RequireAuthenticatedUser();
                    policy.RequireClaim(ClaimTypes.Name, "Snotchie");

                });

                options.AddPolicy("API", policy =>
                {
                    policy.AuthenticationSchemes.Add("Jwt"); // NB: This is used for the Token Based API
                    policy.RequireClaim(ClaimTypes.Role, "Admin");
                });

            });
            #endregion

            services.AddMvc();

            // Ensure all URLS are lowercase
            services.AddRouting(options => options.LowercaseUrls = true);

            #region Services/ INjection Dependencies
            // Add application services.
            services.AddTransient<IEmailSender, AuthMessageSender>();
            services.AddTransient<ISmsSender, AuthMessageSender>();
            services.AddSingleton<IConfiguration>(Configuration);

            services.AddScoped<IBlogRepository, BlogRepository>();

            #endregion
        }

        private SecurityKey getAPISecret()
        {
            return new SymmetricSecurityKey(Encoding.ASCII.GetBytes("Fuck0ff"));
        }

        // This method gets called by the runtime. Use this method to configure the HTTP request pipeline.
        public void Configure(IApplicationBuilder app, IHostingEnvironment env, ILoggerFactory loggerFactory)
        {
            string Authority;

            loggerFactory.AddConsole(Configuration.GetSection("Logging"));
            loggerFactory.AddDebug();

            #region  How Error Codes are displayed
            if (env.IsDevelopment())
            {
                app.UseDeveloperExceptionPage();
                app.UseDatabaseErrorPage();
                app.UseBrowserLink();
            }
            else
            {
                app.UseExceptionHandler("/Home/Error");
            }
            #endregion


            //CORS middleware must precede any defined endpoints in app
            app.UseCors(builder =>
            {
                builder.AllowAnyHeader();
                builder.AllowAnyMethod();
                builder.AllowAnyOrigin();
            });

            app.UseStaticFiles();
            app.UseIdentity();

            //app.UseCookieAuthentication(new CookieAuthenticationOptions
            //{
            //    SlidingExpiration = true,
            //    LoginPath = new PathString("/Account/Login"),
            //    AccessDeniedPath = new PathString("/Account/AccessDenied"),
            //    AuthenticationScheme = "Cookie",
            //});



            #region JWT
            var tokenValidationParameters = new TokenValidationParameters
            {
                // The signing key must match!
                ValidateIssuerSigningKey = true,
                IssuerSigningKey = getAPISecret(),

                //// Validate the JWT Issuer (iss) claim
                ValidateIssuer = false,
                //ValidIssuer = "ExampleIssuer",

                //// Validate the JWT Audience (aud) claim
                ValidateAudience = false,
                //ValidAudience = "ExampleAudience",

                // Validate the token expiry
                ValidateLifetime = true,

                // If you want to allow a certain amount of clock drift, set that here:
                ClockSkew = TimeSpan.Zero
            };



            if (env.IsDevelopment())
            {
                Authority = "http://localhost:9999";
            }
            else
            {
                Authority = "http://productionsite.com";
            }

            //                Microsoft.AspNetCore.Mvc.Internal.ControllerActionInvoker:Warning: Authorization failed for the request at filter 'Microsoft.AspNetCore.Mvc.Authorization.AuthorizeFilter'.
            //Microsoft.AspNetCore.Mvc.ChallengeResult:Information: Executing ChallengeResult with authentication schemes().
            //Microsoft.AspNetCore.Authentication.JwtBearer.JwtBearerMiddleware:Information: AuthenticationScheme: Jwt was challenged.
            //Microsoft.AspNetCore.Authentication.Cookies.CookieAuthenticationMiddleware:Information: AuthenticationScheme: Identity.Application was challenged.

            app.UseJwtBearerAuthentication(new JwtBearerOptions
            {
                RequireHttpsMetadata = false,
                AuthenticationScheme = "Jwt",
                AutomaticAuthenticate = true,
                AutomaticChallenge = true,
                TokenValidationParameters = tokenValidationParameters,
                Authority = Authority,
                //Audience = "http://localhost:9999",
            });

            app.UseOpenIddict();
            #endregion

            app.UseMvc(routes =>
            {
                routes.MapRoute(
                    name: "default",
                    template: "{controller=Home}/{action=Index}/{id?}");
            });
        }
    }
}
```

**Comment by kevinchalet** [comment]
To make them more visible for the community, questions should be posted on StackOverflow: https://stackoverflow.com/questions/tagged/openiddict.

See you there.

**Comment by edslas** [comment]
Bhail, I'm facing the same issue, with openidict for json api, unauthorized access to webapi does not redirect to login page. Did you ever solve this?

**Comment by kinosang** [comment]
@edslas APIs are stateless, they don't do redirect.

**Comment by DBhail** [comment]
@edslas yes...kinduve. I looked at the problem from another angle. Remember with token authentication you have you have a detachment of concerns i.e. a frontend and a backend (API).
The frontend in my case was Angular 4. All I needed to do was check the HTP status code returned. 
400 is when it can't authenticate which actually would be right from a security perspective as you wouldn't want the API to acknowledge its existence otherwise. Then you have 401 for token be invalidated. When my frontend app sees 400 it clears local storage of any authentication and profile info and redirects to a login page. I hope that helps
